### PR TITLE
[OMPIRBuilder] Added `createTeams`

### PR DESCRIPTION
**Pull request by shraiysh**
This patch adds basic support for `omp teams` to the OpenMPIRBuilder.

The outlined function after code extraction is called from a wrapper function with appropriate arguments. This wrapper function is passed to the runtime calls.

This approach is different from the Clang approach - clang directly emits the runtime call to the outlined function. The outlining utility (OutlineInfo) simply outlines the code and generates a function call to the outlined function. After the function has been generated by the outlining utility, there is no easy way to alter the function arguments without meddling with the outlining itself. Hence the wrapper function approach is taken.

### CODE REVIEW

**Comment by shraiysh** [comment]
Apologies for messing up with solving the merge conflicts with HEAD. I have addressed the comments by @kiranchandramohan (from previous comments by @Meinersbur on similar patch).

https://github.com/llvm/llvm-project/pull/65785

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp [hunk 1]
```diff
@@ -5735,6 +5735,140 @@ OpenMPIRBuilder::InsertPointTy OpenMPIRBuilder::createAtomicCompare(
...
+    assert(StaleCI && "Error while outlining - no CallInst user found for the "
+                      "outlined function.");
+    OutlinedFn.addFnAttr(Attribute::AttrKind::AlwaysInline);
+
+    // Create the wrapper function.
+    Builder.SetInsertPoint(StaleCI);
+    SmallVector<Type *> WrapperArgTys{Builder.getPtrTy(), Builder.getPtrTy()};
+    for (auto &Arg : OutlinedFn.args()) {
+      WrapperArgTys.push_back(Arg.getType());
+    }
```
Comment (kiranchandramohan) [inline llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:5818]: Nit: There are a few places where single-statement bodies of loops/single are put in braces.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp [hunk 2]
```diff
@@ -5735,6 +5735,140 @@ OpenMPIRBuilder::InsertPointTy OpenMPIRBuilder::createAtomicCompare(
...
+                           omp::RuntimeFunction::OMPRTL___kmpc_fork_teams),
+                       Args);
+    StaleCI->eraseFromParent();
+
+    // Inlining the outlined teams function in the wrapper. This wrapper is the
+    // argument for the runtime call.
+    assert(OutlinedFn.getNumUses() == 1 &&
+           "More than one use for the outlined function found. Expected only "
+           "one use.");
+    OutlinedFn.addFnAttr(Attribute::AlwaysInline);
```
Comment (kiranchandramohan) [inline llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:5857]: This was done in 5811.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp [hunk 3]
```diff
@@ -5735,6 +5735,140 @@ OpenMPIRBuilder::InsertPointTy OpenMPIRBuilder::createAtomicCompare(
...
+
+    assert(OutlinedFn.getNumUses() == 1 &&
+           "there must be a single user for the outlined function");
+    CallInst *StaleCI = cast<CallInst>(OutlinedFn.user_back());
+    assert(StaleCI && "Error while outlining - no CallInst user found for the "
+                      "outlined function.");
+    OutlinedFn.addFnAttr(Attribute::AttrKind::AlwaysInline);
+
+    // Create the wrapper function.
+    Builder.SetInsertPoint(StaleCI);
```
Comment (kiranchandramohan) [inline llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:5814]: Is this Insertion Point setting required?

**Review by kiranchandramohan — APPROVED** [review]
LGTM. Please wait a couple of days to see whether @jdoerfert or @Meinersbur have comments.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp (hunk 3 above)
Comment (shraiysh) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp]: Not required. Removed it.

**Pull request merged**